android/src/main/java/com/oblador/keychain/KeychainModule.java

```diff
@@ -45,91 +97,245 @@ public KeychainModule(ReactApplicationContext reactContext) {
+            CipherInputStream cipherInputStream = new CipherInputStream(
+                    new ByteArrayInputStream(bytes), cipher);
+
+            ByteArrayOutputStream output = new ByteArrayOutputStream();
+            byte[] buffer = new byte[1024];
+            while (true) {
+                int n = cipherInputStream.read(buffer, 0, buffer.length);
+                if (n <= 0) {
+                    break;
+                }
+                output.write(buffer, 0, n);
+            }
+            return output.toByteArray();
+        } catch (Exception e) {
+            throw new CryptoFailedException("Could not decrypt bytes", e);
+        }
+    }
+
+    private ResultSet getGenericPasswordForOptionsUsingConceal(String service) throws CryptoFailedException {
+        if (!crypto.isAvailable()) {
+            throw new CryptoFailedException("Crypto is missing");
+        }
+
+        service = service == null ? DEFAULT_ALIAS : service;
+
+        byte[] recuser = getBytesFromPrefs(service, LEGACY_DELIMITER + "u");
+        byte[] recpass = getBytesFromPrefs(service, LEGACY_DELIMITER + "p");
+        if (recuser == null || recpass == null) {
+            return null;
+        }
+
+        Entity userentity = Entity.create(KEYCHAIN_DATA + ":" + service + "user");
+        Entity pwentity = Entity.create(KEYCHAIN_DATA + ":" + service + "pass");
+
+        try {
+            byte[] decryptedUsername = crypto.decrypt(recuser, userentity);
+            byte[] decryptedPassword = crypto.decrypt(recpass, pwentity);
+
+            return new ResultSet(service, decryptedUsername, decryptedPassword);
         } catch (Exception e) {
-            Log.e(KEYCHAIN_MODULE, e.getLocalizedMessage());
-            promise.reject(e.getLocalizedMessage(), e);
+            throw new CryptoFailedException("Decryption failed for service " + service, e);
+        }
+    }
+
+    private byte[] getBytesFromPrefs(String service, String prefix) {
+        String key = service + prefix;
+        String value = prefs.getString(service + prefix, null);
+        if (value != null) {
+            return Base64.decode(value, Base64.DEFAULT);
         }
+        return null;
     }
 
     @ReactMethod
     public void resetGenericPasswordForOptions(String service, Promise promise) {
-        service = service == null ? EMPTY_STRING : service;
+        try {
+            resetGenericPasswordForOptionsWithDelimiter(service, DELIMITER, true);
+            promise.resolve(true);
+        } catch (KeyStoreException e) {
+            Log.e(KEYCHAIN_MODULE, e.getMessage());
+            promise.reject(E_UNSUPPORTED_KEYSTORE, e);
+        } catch (KeyStoreAccessException e) {
+            Log.e(KEYCHAIN_MODULE, e.getMessage());
+            promise.reject(E_KEYSTORE_ACCESS_ERROR, e);
+        }
+    }
+
+    private void resetGenericPasswordForOptionsWithDelimiter(String service, String delimiter, boolean resetKeyStore) throws KeyStoreException, KeyStoreAccessException {
```

Comment:
@pcoltau maybe the best would be to have `resetGenericPasswordForOptions` and `resetGenericPasswordForOptionsLegacy` that both accept only one parameter? Then in future versions when conceal is removed it would be easier to refactor the code..